addons/crm/static/src/js/forecast/forecast_service.js

```diff
@@ -0,0 +1,230 @@
+/** @odoo-module alias=crm.ForecastService **/
+import AbstractService from 'web.AbstractService';
+import core from 'web.core';
+
+/**
+ * File intended for `web` (can be used in all kind of forecast configurations)
+ * 
+ * Forecast Period:
+ *   Represents a specific date/time range for a specific model, field and granularity.
+ * 
+ * It is used to add new domain and context constraints related to a specific date/time
+ * field, in order to configure the _read_group_fill_temporal (see core models.py)
+ * method. It will be used when we want to get continuous groups in chronological
+ * order in a specific date/time range.
+ */
+
+/**
+ * Configuration depending on the granularity:
+ * startOf : method to get the start moment of the period from a moment
+ * cycle : amount of 'granularity' periods constituting a cycle
+ * cyclePos : method to get the position (index) in the cycle from a moment (1 is the first index)
+ */
+const GRANULARITY_TABLE = {
+    hour: {
+        startOf: x => x.startOf('hour'),
+        cycle: 24,
+        cyclePos: x => x.hour() + 1,
+    },
+    day: {
+        startOf: x => x.startOf('day'),
+        cycle: 7,
+        cyclePos: x => x.isoWeekday(),
+    },
+    week: {
+        startOf: x => x.startOf('isoWeek'),
+        cycle: 1,
+        cyclePos: x => 1,
+    },
+    month: {
+        startOf: x => x.startOf('month'),
+        cycle: 12,
+        cyclePos: x => x.month() + 1,
+    },
+    quarter: {
+        startOf: x => x.startOf('quarter'),
+        cycle: 4,
+        cyclePos: x => x.quarter(),
+    },
+    year: {
+        startOf: x => x.startOf('year'),
+        cycle: 1,
+        cyclePos: x => 1,
+    },
+};
+
+class ForecastPeriod {
+    /**
+     * This constructor is meant to be used only by the ForecastService (see below)
+     *
+     * @param {string} modelName
+     * @param {string} fieldName
+     * @param {string} granularity can either be : hour, day, week, month, quarter, year
+     * @param {integer} min_groups minimum amount of groups to display, regardless of other constraints
+     */
+    constructor(modelName, fieldName, fieldType, granularity, min_groups) {
```

Comment:
For each variable coming from (and going to) the context, I kept the original syntax with `_` but maybe this is not a good idea ? I did this because changing the name could confuse the issue ?

Reply:
Well we have that issue everywhere we have backend/frontend communication, variables change their names according to the styling. Indeed it makes grepping a bit harder, but min_groups should become minGroups (minGroupsCount is not necessary to avoid replacing all occurences of it).